packages/datadog-plugin-sharedb/test/index.spec.js

```diff
@@ -0,0 +1,194 @@
+'use strict'
+
+const agent = require('../../dd-trace/test/plugins/agent')
+const plugin = require('../src')
+
+wrapIt()
+
+describe('Plugin', () => {
+  let ShareDB
+
+  describe('sharedb', () => {
+    withVersions(plugin, 'sharedb', version => {
+      beforeEach(() => {
+        require('../../dd-trace')
+      })
+
+      describe('without configuration', () => {
+        let backend
+        let connection
+
+        before(() => {
+          return agent.load('sharedb')
+        })
+
+        after(() => {
+          return agent.close()
+        })
+
+        beforeEach(() => {
+          ShareDB = require(`../../../versions/sharedb@${version}`).get()
+
+          backend = new ShareDB({ presence: true })
+          connection = backend.connect()
+        })
+
+        afterEach(() => {
+          connection.close()
+        })
+
+        it('should do automatic instrumentation', done => {
+          const doc = connection.get('some-collection', 'some-id')
+
+          doc.fetch(function (err) {
```

Comment:
Just so I can better review the actual logic and not just the code, can you explain a little bit how this works? For example, this looks like it's only a client, yet the `receive` and `reply` events seem to indicate that it's a server that is instrumented. Is `connection` basically a client to `backend`? If so, right now they are both in the same process, but what happens for remote backends? How is the tracing context propagated to the backend receiving the request on the other end so that it's part of the same distributed trace as the one started by the client?

Reply:
Sure, happy to explain how this works, will see about adding some comments too.

Basically the "backend" object is our server, and the "connection" is the client. When we call `backend.connect()` we are creating a `StreamSocket` that communication is done over: https://github.com/share/sharedb/blob/master/lib/backend.js#L107

"How is the tracing context propagated to the backend receiving the request on the other end so that it's part of the same distributed trace as the one started by the client?" -> It's not, we don't support this.

The client is just a web browser.

Reply:
It could however be server to server, and even in the case of a browser, we support distributed tracing to a server with RUM so it would still result in a disconnected trace.

Reply:
Well in both cases, it'd be over http or websockets, so wouldn't the protocol propagate the tracing information?

Reply:
For HTTP yes, but not for WS since it's too low level and doesn't provide any kind of headers/attributes to attach the tracing context to.